### PR TITLE
OPDS catalog: add book info button

### PR DESCRIPTION
**Pull request by hius07**
In accordance with
https://specs.opds.io/opds-1.2.html#513-summary-and-content
the `content` OPDS catalog entry is shown with this new button.

Useful before download.
Screenshots from flibusta.

--

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/7767)

### CODE REVIEW

**Comment by NiLuJe** [comment]
I'm mildly intrigued how you can end up with a `content` block given:

https://github.com/koreader/koreader/blob/b8d658b922404974ac5acd20a8843e665133d3c3/plugins/opds.koplugin/opdsparser.lua#L75-L78

^^

**Comment by NiLuJe** [comment]
And if we don't, the first calibre book I tried:

```
./luajit: frontend/util.lua:1112: attempt to call method 'gsub' (a nil value)
stack traceback:
        frontend/util.lua:1112: in function 'htmlToPlainTextIfHtml'
        plugins/opds.koplugin/opdsbrowser.lua:678: in function 'callback'
        frontend/ui/widget/button.lua:347: in function 'handleEvent'
        frontend/ui/widget/container/inputcontainer.lua:263: in function 'handleEvent'
        frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
        frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
        frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
        frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
        frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
        frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
        frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
        ...
        frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
        frontend/ui/widget/container/widgetcontainer.lua:95: in function 'propagateEvent'
        frontend/ui/widget/container/widgetcontainer.lua:113: in function 'handleEvent'
        frontend/ui/uimanager.lua:1111: in function 'sendEvent'
        frontend/ui/uimanager.lua:223: in function '__default__'
        frontend/ui/uimanager.lua:1573: in function 'handleInputEvent'
        frontend/ui/uimanager.lua:1653: in function 'handleInput'
        frontend/ui/uimanager.lua:1700: in function 'run'
        ./reader.lua:323: in main chunk
        [C]: at 0x00014161
```

**Comment by NiLuJe** [comment]
Reason: it's a table, not a string.

And even if you flatten it, it tends to crash in fund and interesting ways.? (Plus, potentially unsanitized HTML, brrr!).

In short: nice idea in theory, extremely problematic in practice ;).

**Comment by hius07** [comment]
Flibusta gives in opds output `<content type="text/html">`, so it remains in `item.content` and looks good.
Can this be kept in some way?

**Comment by Frenzie** [comment]
Since it doesn't need a separate request, maybe you could just include a description more directly? (up to few hundred chars and/or a few lines)

**Comment by NiLuJe** [comment]
That doesn't make sense: it's *never* supposed to be a string, it's a parsed tree from luxl:

```
05/27/21-18:56:00 DEBUG item.content {
    ["type"] = "xhtml",
    ["div"] = {
        ["xmlns"] = "http://www.w3.org/1999/xhtml",
        ["p"] = "Rights: Public domain in the USA.",
    },
}
```

(Which means flattening that in a meaningful way is much easier said than done anyway ^^).

**Comment by hius07** [comment]
> include a description more directly?

Excuse me, added to the first post.

Isn't it extremely cruel to kill the whole `content` for all sites just because of calibre?

**Comment by NiLuJe** [comment]
It's not because of calibre, it's because of luxl ;).

It's a block that has nothing relevant in in 75% of the time (ironically, except in calibre's case), but that was responsible for 99% of the "oops, parsing failed and *silently* prevented *some* items from showing up, which is very much worse than whatever we could salvage it for ;).

**Comment by NiLuJe** [comment]
TL;DR: Hard veto from myself unless we also get a working XML & HTML parser ;)

**Comment by Frenzie** [comment]
@hius07 
> Excuse me, added to the first post.

I meant that if it's just 3-5 lines you could probably just show it on download (with an ellipsis for longer in combination with your popup) but that wasn't taking anything @NiLuJe said into account.

**Comment by NiLuJe** [comment]
I don't rightly recall exactly how luxl(mis)behaves, so I wouldn't mind a glance at the actual code of the OPDS feed you showed in your screenshot, but, it's *possible* `<content>blah</blah>` ends up stored as a string if there are no tags inside.

In which case:

```
enabled = item.content ~= nil and type(item.content) == "string",
```

I'd *really* like to see the code of your feed to confirm first, though ;).

**Comment by NiLuJe** [comment]
Mkay, `luxl` is weird, but, yeah, that works, for some reason.

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 1]
```diff
@@ -667,7 +667,18 @@ function OPDSBrowser:showDownloads(item)
                     end,
                 }:chooseDir()
             end,
-        }
+        },
+        {
+            text = _("Book info"),
+            enabled = item.content ~= nil,
```
Comment (NiLuJe) [inline plugins/opds.koplugin/opdsbrowser.lua:673]:
```suggestion
            enabled = item.content ~= nil and type(item.content) == "string",
```

**Review by NiLuJe — CHANGES_REQUESTED** [review]
Until something else blows up, this would be fine by me and should cover at least not horribly breaking the default list of servers.

File: plugins/opds.koplugin/opdsbrowser.lua (hunk 1 above)
Comment (hius07) [reply plugins/opds.koplugin/opdsbrowser.lua]: Thanks, that works with flibusta.

The output from the first post screenshot

`<content type="text/html">
    &lt;p class=book&gt;Трагедия «Гамлет» является одной из вершин творчества Шекспира. В основе пьесы лежит трагическая история датского принца Гамлета, притворившегося безумным, чтобы отомстить убийце отца, завладевшего престолом. Внутренняя душевная борьба, связанная с ужасным открытием тайны смерти отца, в сочетании с неприятием низменной среды королевского двора и желанием исправить мир приводит Гамлета к страданиям, которые становятся причиной его собственной гибели и смерти окружающих его людей.&lt;/p&gt;
   &lt;br/&gt;Перевод: Пастернак Борис Леонидович&lt;br/&gt;Год издания: 1600&lt;br/&gt;Формат: fb2&lt;br/&gt;Язык: ru&lt;br/&gt;Размер: 640 Kb&lt;br/&gt;Скачиваний: 25170&lt;br/&gt;</content>`

**Comment by NiLuJe** [comment]
> Possibly because Flibusta uses html entities for `<` & `>` inside the content block :?.

Yep, that's the ticket.

Which means I can probably salvage that block in a more generic manner...

**Review by NiLuJe — APPROVED** [review]

**Comment by NiLuJe** [comment]
Actually, posting the screenshot in #7768: shouldn't we use the book's title as the widget's title here, instead of "book description:"?

**Comment by hius07** [comment]
Much better!

**Comment by hius07** [comment]
Long title

**Comment by NiLuJe** [comment]
Another minor tweak, because `OPDSBrowser:genItemTableFromCatalog` already takes care of sanitizing this in case tha tuhor field is missing:

```diff
diff --git a/plugins/opds.koplugin/opdsbrowser.lua b/plugins/opds.koplugin/opdsbrowser.lua
index eae36a2ce..0bc643e0e 100644
--- a/plugins/opds.koplugin/opdsbrowser.lua
+++ b/plugins/opds.koplugin/opdsbrowser.lua
@@ -536,10 +536,7 @@ function OPDSBrowser:downloadFile(item, filetype, remote_url)
     -- Download to user selected folder or last opened folder.
     local download_dir = self.getCurrentDownloadDir()
 
-    local filename = item.title .. "." .. filetype
-    if item.author then
-        filename = item.author .. " - " .. filename
-    end
+    local filename = item.text .. "." .. filetype
 
     filename = util.getSafeFilename(filename, download_dir)
     local local_path = download_dir .. "/" .. filename
@@ -674,7 +671,7 @@ function OPDSBrowser:showDownloads(item)
             callback = function()
                 local TextViewer = require("ui/widget/textviewer")
                 UIManager:show(TextViewer:new{
-                    title = item.author .. " - " .. item.title,
+                    title = item.text,
                     text = util.htmlToPlainTextIfHtml(item.content),
                 })
             end,
```

**Comment by hius07** [comment]
When you wrote the comment, I was looking at my Kindle testing `item.text` )

**Comment by hius07** [comment]
Can we change
`enabled = item.content ~= nil and type(item.content) == "string"`
to just
`enabled = type(item.content) == "string"`?

**Comment by hius07** [comment]
> Another minor tweak

Book name in the catalog wasn't equal to the saved file name.
https://github.com/koreader/koreader/blob/b8d658b922404974ac5acd20a8843e665133d3c3/plugins/opds.koplugin/opdsbrowser.lua#L488
https://github.com/koreader/koreader/blob/b8d658b922404974ac5acd20a8843e665133d3c3/plugins/opds.koplugin/opdsbrowser.lua#L539-L541
Should we keep the difference? Title or author to be the first?

**Comment by NiLuJe** [comment]
> Can we change
> `enabled = item.content ~= nil and type(item.content) == "string"`
> to just
> `enabled = type(item.content) == "string"`?

Yeah, good call (`type(nil)` returns `"nil"`).

**Comment by NiLuJe** [comment]
> > Another minor tweak
> 
> Book name in the catalog wasn't equal to the saved file name.
> 
> https://github.com/koreader/koreader/blob/b8d658b922404974ac5acd20a8843e665133d3c3/plugins/opds.koplugin/opdsbrowser.lua#L488
> 
> https://github.com/koreader/koreader/blob/b8d658b922404974ac5acd20a8843e665133d3c3/plugins/opds.koplugin/opdsbrowser.lua#L539-L541
> 
> Should we keep the difference? Title or author to be the first?
> 
> As for my use case, I'd keep the difference.
> When browsing opds catalog the books by one author should be sorted by the title.
> Downloaded books by various authors might be mixed in a folder, so the filenames should begin with the author.

I hadn't really thought through the differences, just noticed 'em and thought, "oh, well, DRYer".

But, yeah, everything you said makes perfect sense, agreed!

**Comment by hius07** [comment]
Seems too much of the bold text?

**Review by NiLuJe — APPROVED** [review]

File: plugins/opds.koplugin/opdsbrowser.lua [hunk 2]
```diff
@@ -667,7 +668,18 @@ function OPDSBrowser:showDownloads(item)
                     end,
                 }:chooseDir()
             end,
-        }
+        },
+        {
+            text = _("Book info"),
```
Comment (NiLuJe) [inline plugins/opds.koplugin/opdsbrowser.lua:673]: Possibly make it `Book information`, to match the hold menu in the FM?
